build/downloader.ts

```diff
@@ -0,0 +1,110 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as releases from '@hashicorp/js-releases';
+
+function getPlatform(platform: string) {
+  if (platform === 'win32') {
+    return 'windows';
+  }
+  if (platform === 'sunos') {
+    return 'solaris';
+  }
+  return platform;
+}
+
+function getArch(arch: string) {
+  // platform | terraform-ls  | extension platform | vs code editor
+  //    --    |           --  |         --         | --
+  // macOS    | darwin_amd64  | darwin_x64         | ✅
+  // macOS    | darwin_arm64  | darwin_arm64       | ✅
+  // Linux    | linux_amd64   | linux_x64          | ✅
+  // Linux    | linux_arm     | linux_armhf        | ✅
+  // Linux    | linux_arm64   | linux_arm64        | ✅
+  // Windows  | windows_386   | win32_ia32         | ✅
+  // Windows  | windows_amd64 | win32_x64          | ✅
+  // Windows  | windows_arm64 | win32_arm64        | ✅
+  if (arch === 'ia32') {
+    return '386';
+  }
+  if (arch === 'x64') {
+    return 'amd64';
+  }
+  if (arch === 'armhf') {
+    return 'arm';
+  }
+  return arch;
+}
+
+interface ExtensionInfo {
+  extensionVersion: string;
+  languageServerVersion: string;
+}
+
+function getExtensionInfo(): ExtensionInfo {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const pjson = require('../package.json');
+  return {
+    extensionVersion: pjson.version,
+    languageServerVersion: pjson.langServer.version,
+  };
+}
+
+async function run(platform: string, architecture: string) {
+  const cwd = path.resolve(__dirname);
+
+  const buildDir = path.basename(cwd);
+  const repoDir = cwd.replace(buildDir, '');
+  const installPath = path.join(repoDir, 'bin');
+  if (fs.existsSync(installPath)) {
+    console.log('terraform-ls path exists. Exiting');
+    return;
+  }
+
+  fs.mkdirSync(installPath);
+
+  const extInfo = getExtensionInfo();
+  console.log(extInfo);
+
+  const userAgent = `Terraform-VSCode/${extInfo.extensionVersion}`;
```

Comment:
I see the new User-Agent omits VS Code version, I'm guessing this was intentional? I suppose that the download stats won't really be as useful after this goes live anyway - i.e. I suppose we don't care as much about downloads from CI 😄  - so I have no strong opinions here - just checking your intentions.

Reply:
It was intentional in the sense that at the time I was more interested in getting the script working than reporting which vscode version we're using. Since this is now outside of vscode, there isn't an easy variable to pull to get the vscode version, I would have to write something extra to find that out.

As you point out, this is less relevant as we're building in CI and that there isn't a 'VS Code' there to track, however it might be a good idea to put something there so we can filter CI runs from the data so that we can see those coming from other editors. Ideas for a name/version to use here?

Reply:
Updated with modified user agent